File: dev/xcheck/Rng.java
import java.io.*;
public class Rng {
    public static void main(String[] a) throws Exception {
        int st = 12345;
        BufferedReader r = new BufferedReader(new FileReader("rng.txt"));
        long mism = 0; String line; int i=0;
        while((line=r.readLine())!=null){
            st = st + 0x6D2B79F5;               // wraps like |0
            int t = (st ^ (st >>> 15)) * (1 | st);   // Math.imul == int multiply
            t = (t + ((t ^ (t >>> 7)) * (61 | t))) ^ t;
            double v = ((long)((t ^ (t >>> 14))) & 0xFFFFFFFFL) / 4294967296.0; // >>>0 then /2^32
            float f = (float)(v * 3.7 - 1.85);
            String[] p = line.split(" ");
            if (Double.doubleToRawLongBits(v) != Long.parseUnsignedLong(p[0],16)) mism++;
            if (Float.floatToRawIntBits(f) != (int)Long.parseLong(p[1],16)) mism++;
            i++;
        }
        System.out.println("Java mulberry32+f32: draws="+i+" mismatches="+mism);
    }
}

File: dev/xcheck/rng.rs
use std::fs;
fn main(){
    let data = fs::read_to_string("rng.txt").unwrap();
    let mut st: i32 = 12345;
    let (mut n, mut mism) = (0u64, 0u64);
    for line in data.lines(){
        st = st.wrapping_add(0x6D2B79F5u32 as i32);
        let mut t = (st ^ ((st as u32) >> 15) as i32).wrapping_mul(1 | st);
        t = (t.wrapping_add((t ^ ((t as u32) >> 7) as i32).wrapping_mul(61 | t))) ^ t;
        let v = ((t ^ ((t as u32) >> 14) as i32) as u32) as f64 / 4294967296.0;
        let f = (v * 3.7 - 1.85) as f32;
        let p: Vec<&str> = line.split(' ').collect();
        if v.to_bits() != u64::from_str_radix(p[0],16).unwrap() { mism+=1; }
        if f.to_bits() != u32::from_str_radix(p[1],16).unwrap() { mism+=1; }
        n+=1;
    }
    println!("Rust mulberry32+f32: draws={} mismatches={}", n, mism);
}
